### PR TITLE
samples: Bluetooth: hci_uart: Implement NOP Command Complete

### LINKED ISSUE #15333: hci_uart controller driver loses sync after host driver is reset
The `hci_uart` controller driver will lose synchronization with the host, if the host is reset while sending data. If the controller driver is reset, the host can wait for a `NOP` from the controller before continuing;
this can serve as a synchronization mechanism. However, the `hci_uart` sample does not emit a NOP command.

Solution:
Emit a `NOP` command from the `hci_uart` controller driver,
to signal to the host that the controller is ready.

### PR DESCRIPTION
**Pull request by carlescufi**
Implement the Controller counterpart to CONFIG_BT_WAIT_NOP so that it
issues a NOP Command Complete event after booting up, to signal to the
Host that it is ready to receive HCI traffic.

Fixes #15333

Signed-off-by: Carles Cufi <carles.cufi@nordicsemi.no>

### CODE REVIEW

**Comment by carlescufi** [comment]
Output of a reset with this patch:

image

**Comment by lemrey** [comment]
This should address #15333.

File: samples/bluetooth/hci_uart/src/main.c [hunk 1]
```diff
@@ -357,6 +357,32 @@ void main(void)
...
+			const struct bt_hci_evt_cmd_complete cc;
+		} __packed cc_evt = {
+			.h4 = H4_EVT,
+			.hdr = {
+				.evt = BT_HCI_EVT_CMD_COMPLETE,
+				.len = sizeof(struct bt_hci_evt_cmd_complete),
+			},
+			.cc = {
+				.ncmd = 1,
+				.opcode = BT_OP_NOP,
```
Comment (jhedberg) [inline samples/bluetooth/hci_uart/src/main.c]: I know that it makes no difference in practice, however this should really be sys_cpu_to_le16(BT_OP_NOP) to keep track of what endianness `.opcode` is.

**Review by jhedberg — CHANGES_REQUESTED** [review]
Looks generally good, except for one minor endianness "correctness" issue

File: samples/bluetooth/hci_uart/src/main.c (hunk 1 above)
Comment (carlescufi) [reply samples/bluetooth/hci_uart/src/main.c]: Is this actually allowed on an initializer?

Comment (jhedberg) [reply samples/bluetooth/hci_uart/src/main.c]: On big endian this maps to `__bswap_16()` which is defined as follows:

```
#define __bswap_16(x) ((u16_t) ((((x) >> 8) & 0xff) | (((x) & 0xff) << 8))) 
```

I'd think the compiler should be able to handle that even for the initialiser of a const variable, but you could always try it by using `__bswap_16(BT_OP_NOP)`

Comment (carlescufi) [reply samples/bluetooth/hci_uart/src/main.c]: It is. Updated, thanks!

**Review by jhedberg — APPROVED** [review]

File: samples/bluetooth/hci_uart/src/main.c [hunk 2]
```diff
@@ -357,6 +357,32 @@ void main(void)
...
+				.len = sizeof(struct bt_hci_evt_cmd_complete),
+			},
+			.cc = {
+				.ncmd = 1,
+				.opcode = sys_cpu_to_le16(BT_OP_NOP),
+			},
+		};
+
+		for (i = 0; i < sizeof(cc_evt); i++) {
+			uart_poll_out(hci_uart_dev, *(((u8_t *)&cc_evt)+i));
```
Comment (anangl) [inline samples/bluetooth/hci_uart/src/main.c]: `(const u8_t *)`, just for consistency

**Review by anangl — APPROVED** [review]

**Review by joerchan — APPROVED** [review]

File: samples/bluetooth/hci_uart/src/main.c (hunk 2 above)
Comment (carlescufi) [reply samples/bluetooth/hci_uart/src/main.c]: Fixed, thank you

**Review by lemrey — APPROVED** [review]
LGTM

**Pull request merged**